Helmchart: Provision Standalone Dag Processor

gmsantos opened the pull request:
Initial support to deploy standalone dag processor from Helm on Airflow 2.3+

It is disabled by default for now.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

gmsantos committed:
Add Standalone Dag Processor to Helm

Add Dag Processor templates and values to schema, at the moment
the standalone dag processor is disabled by default.

gmsantos committed:
Control standalone_dag_processor config based on chart values

gmsantos committed:
Add tests for Dag Processor

gmsantos commented on chart/templates/dagprocessor/dag-processor-deployment.yaml:
```diff
@@ -0,0 +1,214 @@
+  name: {{ .Release.Name }}-dag-processor
+  labels:
+    tier: airflow
+    component: dag-processor
```
This is the first component in templates with a composed name. How should it be referenced?

At this moment, I'm following the pattern:

`dagprocessor` as the path name inside the template folder
`dag-processor` in labels and in the k8s object name
`dagProcessor` in the values file

gmsantos replied:
A more consistent alternative would be:

`dag-processor` as the path inside the template folder, labels and the k8s object name
`dagProcessor` in the values file

jedcunningham replied:
That second one sounds good to me.

gmsantos commented on chart/templates/dagprocessor/dag-processor-serviceaccount.yaml:
```diff
@@ -0,0 +1,41 @@
+# under the License.
+
+################################
+## Airflow Triggerer ServiceAccount
```
```suggestion
## Airflow Dag Processor ServiceAccount
```

gmsantos commented on chart/values.schema.json:
```diff
@@ -2101,6 +2102,264 @@
                 }
             }
         },
+        "dagProcessor": {
```
This block is based on the triggerer component.

gmsantos commented on chart/templates/_helpers.yaml:
```diff
@@ -653,6 +664,14 @@ Create the name of the cleanup service account to use
+  - -c
+  - |
+    CONNECTION_CHECK_MAX_COUNT=0 AIRFLOW__LOGGING__LOGGING_LEVEL=ERROR exec /entrypoint \
+    airflow jobs check --job-type SchedulerJob --hostname $(hostname)
```
We don't have a dedicated jobs check command for standalone dag processor for now. I can create in this same PR or in a separate one.

jedcunningham replied:
In the meantime, you can probably leave off the job type and it'll still work.

gmsantos replied:
Makes sense

gmsantos commented on tests/charts/test_basic_helm_chart.py:
```diff
@@ -97,6 +97,78 @@ def test_basic_deployments(self):
+                },
+                'labels': {"TEST-LABEL": "TEST-VALUE"},
+                "fullnameOverride": "TEST-BASIC",
+                "dagProcessor": {'enabled': True},
```
If we keep it enabled by default, I can merge the assertions into the test_basic_deployment

jedcunningham replied:
I think being disabled by default at this point makes sense.

gmsantos commented on tests/charts/test_dag_processor.py:
```diff
@@ -0,0 +1,431 @@
+        docs = render_chart(
+            values={
+                "dagProcessor": {
+                    "enabled": True,
```
Same here, if we enable dagProcessor by default, we can remove it from the test values

gmsantos committed:
Create service account only if the service is enabled

gmsantos committed:
Use ternary to check for multiNamespaceMode config

jedcunningham reviewed (commented):
We no longer need the DAGs mounted in the scheduler if this is enabled, right?

gmsantos committed:
Move dag mount from scheduler when dag processor is enabled

potiuk reviewed (commented):
Cautious LGTM. But let's wait for the Helm Gurus.

jedcunningham reviewed (commented):
This all looks good to me, but someone needs to test this when DAGs aren't available in the scheduler (gitsync or PVC), and make sure DAG level callbacks work too. I'm happy to do it once I finish my summit talk prep.

gmsantos committed:
Mount DAG folder to scheduler if it's a local executor

With local executor, the scheduler is the worker and it
needs access to the DAGs

gmsantos committed:
Consider LocalKubernetesOperator as a local executor

gmsantos committed:
Fix condition to include dags mount on scheduler

see the explanation for this logic https://github.com/apache/airflow/pull/23711#issuecomment-1165007592

jedcunningham commented on chart/templates/scheduler/scheduler-deployment.yaml:
```diff
@@ -141,7 +141,7 @@ spec:
           {{- include "standard_airflow_environment" . | indent 10 }}
         {{- end }}
-        {{- if .Values.dags.gitSync.enabled }}
+        {{- if and (not .Values.dagProcessor.enabled) .Values.dags.gitSync.enabled }}
```
Don't we still need this if we have dagProcessor and local?

gmsantos replied:
Makes sense, local executor still requires DAGs to be mounted.

jedcunningham commented on chart/templates/scheduler/scheduler-deployment.yaml:
```diff
@@ -202,13 +202,13 @@ spec:
 {{ toYaml .Values.scheduler.extraVolumeMounts | indent 12 }}
 {{- end }}
-{{- if .Values.dags.gitSync.enabled }}
+{{- if and (not .Values.dagProcessor.enabled) .Values.dags.gitSync.enabled }}
```
Same here.

jedcunningham reviewed (commented):
It'd be good to get more comprehensive test coverage for the various combos of dag processor, gitsync, persistence, and the dag mount/gitsync containers.

gmsantos committed:
Refactor condition to skip dag mounts or gitsync on scheduler

gmsantos commented on tests/charts/test_basic_helm_chart.py:
```diff
@@ -113,6 +113,84 @@ def test_basic_deployments(self, version):
             ), f"Missing label TEST-LABEL on {k8s_name}. Current labels: {labels}"
 
+    @parameterized.expand(["2.3.2", "2.4.0", "default"])
+    def test_basic_deployment_with_standalone_scheduler(self, version):
```
```suggestion
    def test_basic_deployment_with_standalone_dag_processor(self, version):
```

gmsantos committed:
Rename test

gmsantos committed:
Merge branch 'main' into helm-dag-processor

gmsantos committed:
Merge branch 'main' into helm-dag-processor

gmsantos committed:
Merge branch 'main' into helm-dag-processor

jedcunningham committed:
Merge branch 'main' into helm-dag-processor
